// htmlbars/dist/cjs/simple-html-tokenizer/tokens.js
"use strict";
function StartTag(tagName, attributes, selfClosing) {
  this.type = 'StartTag';
  this.tagName = tagName || '';
  this.attributes = attributes || [];
  this.selfClosing = selfClosing === true;
  this._currentAttribute = undefined;
}

exports.StartTag = StartTag;StartTag.prototype = {
  addToTagName: function(char) {
    this.tagName += char;
  },

  startAttribute: function(char) {
    this._currentAttribute = [char.toLowerCase(), "", null];
    this.attributes.push(this._currentAttribute);
  },

  addToAttributeName: function(char) {
    this._currentAttribute[0] += char;
  },

  markAttributeQuoted: function(value) {
    this._currentAttribute[2] = value;
  },

  addToAttributeValue: function(char) {
    this._currentAttribute[1] = this._currentAttribute[1] || "";
    this._currentAttribute[1] += char;
  },

  finalizeAttributeValue: function() {
    if (this._currentAttribute) {
      if (this._currentAttribute[2] === null) {
        this._currentAttribute[2] = false;
      }
      this._currentAttribute = undefined;
    }
  },

  finalize: function() {
    this.finalizeAttributeValue();
    return this;
  }
};

function EndTag(tagName) {
  this.type = 'EndTag';
  this.tagName = tagName || '';
}

exports.EndTag = EndTag;EndTag.prototype = {
  addToTagName: function(char) {
    this.tagName += char;
  },
  finalize: function () {
    return this;
  }
};

function Chars(chars) {
  this.type = 'Chars';
  this.chars = chars || "";
}

exports.Chars = Chars;Chars.prototype = {
  addChar: function(char) {
    this.chars += char;
  }
};

function Comment(chars) {
  this.type = 'Comment';
  this.chars = chars || '';
}

exports.Comment = Comment;Comment.prototype = {
  addChar: function(char) {
    this.chars += char;
  },

  finalize: function() { return this; }
};
